src/Sulu/Bundle/EventLogBundle/Application/Collector/DomainEventCollector.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\EventLogBundle\Application\Collector;
+
+use Doctrine\Common\EventSubscriber;
+use Doctrine\ORM\Event\OnClearEventArgs;
+use Doctrine\ORM\Event\PostFlushEventArgs;
+use Doctrine\ORM\Events;
+use Sulu\Bundle\EventLogBundle\Application\Dispatcher\DomainEventDispatcherInterface;
+use Sulu\Bundle\EventLogBundle\Domain\Event\DomainEvent;
+
+class DomainEventCollector implements DomainEventCollectorInterface, EventSubscriber
+{
+    /**
+     * @var DomainEventDispatcherInterface
+     */
+    private $domainEventDispatcher;
+
+    /**
+     * @var DomainEvent[]
+     */
+    private $eventsToBeDispatched = [];
+
+    public function __construct(
+        DomainEventDispatcherInterface $domainEventDispatcher
+    ) {
+        $this->domainEventDispatcher = $domainEventDispatcher;
+    }
+
+    public function getSubscribedEvents()
+    {
+        return [
+            Events::onClear,
+            Events::postFlush,
+        ];
+    }
+
+    public function collect(DomainEvent $domainEvent): void
+    {
+        $this->eventsToBeDispatched[] = $domainEvent;
+    }
+
+    public function onClear(OnClearEventArgs $args): void
+    {
+        $this->eventsToBeDispatched = [];
+    }
+
+    public function postFlush(PostFlushEventArgs $args): void
+    {
+        $batchIdentifier = \uniqid('', true);
+        $batchEvents = $this->eventsToBeDispatched;
+
+        $this->eventsToBeDispatched = [];
```

Comment:
should we clear the array before it was dispatched, something like would be more fail safe:

```php
foreach ($this->eventsToBeDispatched as $key => $domainEvent) {
    if (!$domainEvent->getEventBatch()) {
        $domainEvent->setEventBatch($batchIdentifier);
    }

    $this->domainEventDispatcher->dispatch($domainEvent);

    unset($this->eventsToBeDispatched[$key]); // remove after really dispatched
}
```

Reply:
i actually tried something like this initially, but the problem is that the `postFlush` method will be called again when an event is dispatched because the `EventRecord` is flushed to the database. because of that, events will be emitted multiple times if we dont remove them from the `eventsToBeDispatched`